# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: xpcom/threads/nsThread.h
# Commit: d04f7a7ec375
# Full Hash: d04f7a7ec375bc4bf08b39875f86608a8920691e
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2021-04-07 03:19:44
# Regressor Bug: 1695580
# File Overlap Count: 1
# Description:
#   Bug 1695580 - In xpcom, cancel pending DelayedRunnable timers on shutdown. r=KrisWright
#   
#   Because DelayedRunnables are fire-and-forget, there is no way for a targeted
#   EventTarget to clean them up on shutdown. Thus if a timer fires after
#   EventTarget shutdown it will fail to dispatch the timer event, and avoid
# ==============================================================================

diff -r 2103cd9e58b7 -r d04f7a7ec375 xpcom/threads/nsThread.h
--- a/xpcom/threads/nsThread.h	Tue Apr 06 12:16:11 2021 +0000
+++ b/xpcom/threads/nsThread.h	Tue Apr 06 12:16:11 2021 +0000
@@ -21,6 +21,7 @@
 #include "mozilla/TaskDispatcher.h"
 #include "mozilla/TimeStamp.h"
 #include "mozilla/UniquePtr.h"
+#include "nsIDelayedRunnableObserver.h"
 #include "nsIDirectTaskDispatcher.h"
 #include "nsIEventTarget.h"
 #include "nsISerialEventTarget.h"
@@ -31,6 +32,7 @@
 
 namespace mozilla {
 class CycleCollectedJSContext;
+class DelayedRunnable;
 class SynchronizedEventQueue;
 class ThreadEventQueue;
 class ThreadEventTarget;
@@ -152,6 +154,7 @@
 // A native thread
 class nsThread : public nsIThreadInternal,
                  public nsISupportsPriority,
+                 public nsIDelayedRunnableObserver,
                  public nsIDirectTaskDispatcher,
                  private mozilla::LinkedListElement<nsThread> {
   friend mozilla::LinkedList<nsThread>;
@@ -258,6 +261,10 @@
     mUseHangMonitor = aValue;
   }
 
+  void OnDelayedRunnableCreated(mozilla::DelayedRunnable* aRunnable) override;
+  void OnDelayedRunnableScheduled(mozilla::DelayedRunnable* aRunnable) override;
+  void OnDelayedRunnableRan(mozilla::DelayedRunnable* aRunnable) override;
+
  private:
   void DoMainThreadSpecificProcessing() const;
 